42.md

```diff
@@ -0,0 +1,44 @@
+NIP-42
+======
+
+External Embed Sizing
+---------------
+
+`draft` `optional` `author:deanpress`
+
+An `embed-sizes` array in notes that allows relays to store information about the filesizes of external embeddable media (i.e. URLs to images and other media in posts and avatars).
+
+Clients can auto-embed small media within a configured limit, and not auto-embed larger media by default (and give the user an option to click "View Embedded Media"), to save user bandwidth and prevent abuse.
+
+Spec
+---------------
+```
+tag: embed-sizes
+values:
+ - [array of strings of filesizes in bytes, in order of urls in note content]: required
+```
+
+Example
+---------------
+```json
+{
+    "pubkey": "<pub-key>",
+    "created_at": 1000000000,
+    "kind": 1,
+    "tags": [
+      ["embed-sizes", ["500000", "1500000", "0"]]
+    ],
+    "content": "https://imgur.com/1.png and https://imgur.com/2.png and https://imgu2.com/3.png.",
+    "id": "<event-id>"
+}
+```
+In this example the `1.png` embed is 500kb, `2.png` is 1.5MB, and `3.png` was not estimated (due to no `Content-Length` header, or the header doesn't match the real content's size).
+
+Client Behavior
+---------------
+Clients SHOULD check the `embed-sizes` array for each externally linked embeddable media url in a note and only auto-embed the media if it's within the client's (or user's) configured max allowed auto-embeddable filesize. If the `embed-sizes` value of the media embed is `"0"`, also don't auto-embed the media.
```

Comment:
Why does the client trust these values? If the user wishes to avoid excessive data usage, they will need to implement checks of these file size limits even without this embedded data.

Reply:
If a relay deliberately provides false data then they can simply be removed by the user and/or the client as a reliable relay. The idea is that a good integration of this NIP helps with abuse/spam of extremely large external auto-embedded media.